File: Other Data Structures_qstackheap/le146_lruCache.py
"""
Design and implement a data structure for Least Recently Used (LRU) cache. It should support the following operations: get and set.

get(key) - Get the value (will always be positive) of the key if the key exists in the cache, otherwise return -1.
set(key, value) - Set or insert the value if the key is not already present. When the cache reached its capacity, 
it should invalidate the least recently used item before inserting a new item.
"""
class listNode(object):
    def __init__(self, key = None, val = None, next = None):
        self.key = key
        self.val = val
        self.next = next

class LRUCache(object):

    def __init__(self, capacity):
        """
        :type capacity: int
        """
        self.cap = capacity
        self.head = listNode()
        self.tail = self.head
        self.hash = {}
        
    def pushBack(self, node):
        self.tail.next = node
        self.hash[node.key] = self.tail
        self.tail = node
        
    def popFront(self):
        if not self.head.next:  # This can be omitted
            return
        del self.hash[self.head.next.key]
        self.head.next = self.head.next.next
        self.hash[self.head.next.key] = self.head
        
    def kick(self, prev): # move prev.next to the back
        node = prev.next
        if node == self.tail:
            return
        prev.next = node.next
        if node.next is not None:
            self.hash[node.next.key] = prev
        self.pushBack(node)
    
    def get(self, key):
        """
        :rtype: int
        """
        if key in self.hash:
            self.kick(self.hash[key])
            return self.hash[key].next.val
        return -1
        
    def set(self, key, value):
        """
        :type key: int
        :type value: int
        :rtype: nothing
        """
        if key not in self.hash:
            self.pushBack(listNode(key,value,None))
            if len(self.hash) > self.cap:
                 self.popFront()
        else:
            self.kick(self.hash[key])
            self.hash[key].next.val = value

File: Other Data Structures_qstackheap/le146_lruCache.java
public class LRUCache {
    static class Node {
        int val;
        int key;
        Node prev, next;
        Node(int key, int val) {
            this.key = key;
            this.val = val;
        }
    }
    int limit;
    Node head, tail;
    Map<Integer, Node> hmap;
    public LRUCache(int capacity) {
        this.limit = capacity;
        this.hmap = new HashMap<Integer, Node>();    
    }
    
    public int get(int key) {
        Node node = hmap.get(key);
        if (node != null) {
            remove(node);
            moveToHead(node);
            return node.val;
        } 
        return -1;
    }
    
    public void put(int key, int value) {
        // hardest part
        Node node = null;
        if (hmap.containsKey(key)) {
            node = hmap.get(key);
            node.val = value;
            remove(node);
        } else if (hmap.size() < limit){
            node = new Node(key, value);     
        } else {
            node = new Node(key, value);  
            remove(tail);
        }
        moveToHead(node);
    }
    
    private void moveToHead(Node node) {
        hmap.put(node.key, node);
        if (head == null) {
            head = tail = node;
        } else {
            head.prev = node;
            node.next = head;
            head = head.prev;   
        }

    }
    
    private void remove(Node node) {
        hmap.remove(node.key);
        if (node.prev != null) {
            node.prev.next = node.next;
        }
        if (node.next != null) {
            node.next.prev = node.prev;
        }
        
        if (node == head) {
            head = head.next;
        }
        if (node == tail) {
            tail= tail.prev;
        }
        node.next = node.prev = null;
    } 
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */
